provider/adguardhome/adguardhome.go

```diff
@@ -0,0 +1,194 @@
+			adguard.RequestEditorFn(
+				func(ctx context.Context, req *http.Request) error {
+					req.Header.Add("Authorization", basicAuth(adguardHomeUser, adguardHomePass))
+					return nil
+				},
+			),
+			adguard.RequestEditorFn(
+				func(ctx context.Context, req *http.Request) error {
+					req.Header.Add("User-Agent", "ExternalDNS")
+					return nil
+				},
+			),
+		)
+		return nil
+	}))
+
+	if clientOk != nil {
+		return nil, fmt.Errorf("failed to create the adguard home api client")
+	}
+
+	p := &AdguardHomeProvider{
+		client:       *client,
+		domainFilter: domainFilter,
+		DryRun:       dryRun,
+	}
+
+	return p, nil
+}
+
+// ApplyChanges implements Provider, syncing desired state with the AdguardHome server Local DNS.
+func (p *AdguardHomeProvider) ApplyChanges(ctx context.Context, changes *plan.Changes) error {
+	// Get current filter values
+	resp, err := p.client.FilteringStatusWithResponse(ctx)
+	if err != nil {
+		log.Errorf("Error %s", err)
+		return err
+	}
+
+	debugJ, _ := json.MarshalIndent(changes, "", "    ")
+	log.Debugf("changes, %s", string(debugJ))
+	userRules := *resp.JSON200.UserRules
+
+	for _, createValue := range append(changes.Create, changes.UpdateNew...) {
+		newRule := genAdguardCustomRuleRecord(createValue)
+		log.Debugf("add custom rule %s", newRule)
+		userRules = append(userRules, newRule)
+	}
+
+	for _, deleteValue := range append(changes.UpdateOld, changes.Delete...) {
+		userRules = removeRule(userRules, genAdguardCustomRuleRecord(deleteValue), false)
+	}
+
+	apply, err := p.client.FilteringSetRulesWithResponse(ctx, adguard.SetRulesRequest{
+		Rules: &userRules,
+	})
+
+	if apply.StatusCode() != 200 {
+		log.Errorf("Failed to sync records %s", apply.Status())
+	}
+	return err
+}
+
+func removeRule(rules []string, ruleToRemove string, finishAfterFirst bool) []string {
+	userRules := rules
+	for userRuleIndex, userRule := range rules {
+		if userRule == ruleToRemove {
+			log.Debugf("deleted rule %s", ruleToRemove)
+			userRules = append(userRules[:userRuleIndex], userRules[userRuleIndex+1:]...)
+			if finishAfterFirst {
+				return userRules
+			}
+		}
+	}
+	return userRules
+}
+
+// Records implements Provider, populating a slice of endpoints from
+// AdguardHome local DNS.
+func (p *AdguardHomeProvider) Records(ctx context.Context) ([]*endpoint.Endpoint, error) {
+	resp, err := p.client.FilteringStatusWithResponse(ctx)
+	if err != nil {
+		log.Errorf("Error %s", err)
+	}
+
+	log.WithFields(log.Fields{
+		"func":  "Records",
+		"rules": resp.JSON200.UserRules,
+	}).Debugf("retrieved rules")
+	var ret []*endpoint.Endpoint
+	for _, customRule := range *resp.JSON200.UserRules {
+		parsed := parseAdguardCustomRuleRecord(customRule)
+		if parsed != nil {
+			ret = append(ret, parsed)
+		}
+	}
+
+	return ret, nil
+}
+
+func genAdguardCustomRuleRecord(e *endpoint.Endpoint) string {
+	return fmt.Sprintf("%s$dnstype=%s,dnsrewrite=NOERROR;%s;%s", e.DNSName, e.RecordType, e.RecordType, e.Targets.String())
```

Comment:
Suggest wrapping the DNSName in |%s| because otherwise a record for example.com will also match test.example.com or test.example.com.example.com

Default matching mode in AGH appears to be "contains substring" 

As per https://github.com/AdguardTeam/AdGuardHome/wiki/Hosts-Blocklists#adblock-style
wrapping the record name in || to match the beginning and end should cause it to perform an exact match.